monitor/internal/kubernetes/monitor.go

```diff
@@ -0,0 +1,142 @@
+package kubernetesmonitor
+
+import (
+	"context"
+	"fmt"
+
+	"go.uber.org/zap"
+	"k8s.io/client-go/kubernetes"
+	kubecache "k8s.io/client-go/tools/cache"
+
+	"github.com/aporeto-inc/trireme-lib/collector"
+	"github.com/aporeto-inc/trireme-lib/monitor/extractors"
+
+	"github.com/aporeto-inc/trireme-lib/monitor/config"
+	"github.com/aporeto-inc/trireme-lib/monitor/registerer"
+
+	dockermonitor "github.com/aporeto-inc/trireme-lib/monitor/internal/docker"
+)
+
+// KubernetesMonitor implements a monitor that sends pod events upstream
+// It is implemented as a filter on the standard DockerMonitor.
+// It gets all the PU events from the DockerMonitor and if the container is the POD container from Kubernetes,
+// It connects to the Kubernetes API and adds the tags that are coming from Kuberntes that cannot be found
+type KubernetesMonitor struct {
+	dockerMonitor       *dockermonitor.DockerMonitor
+	kubeClient          kubernetes.Interface
+	localNode           string
+	handlers            *config.ProcessorConfig
+	cache               *cache
+	kubernetesExtractor extractors.KubernetesMetadataExtractorType
+
+	podStore          kubecache.Store
+	podController     kubecache.Controller
+	podControllerStop chan struct{}
+
+	enableHostPods bool
+}
+
+// New returns a new kubernetes monitor.
+func New() *KubernetesMonitor {
+	kubeMonitor := &KubernetesMonitor{}
+	kubeMonitor.cache = newCache()
+
+	return kubeMonitor
+}
+
+// SetupConfig provides a configuration to implmentations. Every implmentation
+// can have its own config type.
+func (m *KubernetesMonitor) SetupConfig(registerer registerer.Registerer, cfg interface{}) error {
+
+	defaultConfig := DefaultConfig()
+
+	if cfg == nil {
+		cfg = defaultConfig
+	}
+
+	kubernetesconfig, ok := cfg.(*Config)
+	if !ok {
+		return fmt.Errorf("Invalid configuration specified")
+	}
+
+	kubernetesconfig = SetupDefaultConfig(kubernetesconfig)
+
+	processorConfig := &config.ProcessorConfig{
+		Policy:    m,
+		Collector: collector.NewDefaultCollector(),
+	}
+
+	// As the Kubernetes monitor depends on the DockerMonitor, we setup the Docker monitor first
+	dockerMon := dockermonitor.New()
+	dockerMon.SetupHandlers(processorConfig)
+
+	dockerConfig := dockermonitor.DefaultConfig()
+	dockerConfig.EventMetadataExtractor = kubernetesconfig.DockerExtractor
+
+	// we use the defaultconfig for now
+	if err := dockerMon.SetupConfig(nil, dockerConfig); err != nil {
```

Comment:
This will be an issue with the bigger integrations and the APIs. Isn't this propagated through the instantiation of hte library (I mean the defaultConfig for docker)?